src/NUnitFramework/framework/Internal/Reflect.cs

```diff
@@ -463,5 +463,32 @@ internal static bool IsVoidOrUnit(Type type)
 
             return type == typeof(void) || type.FullName == "Microsoft.FSharp.Core.Unit";
         }
+
+        /// <summary>
+        /// Returns the property information for a given named indexer.
+        /// </summary>
+        /// <param name="type">Type to reflect on for indexer.</param>
+        /// <param name="indexerTypes">List of indexer types that matches the indexer type order.</param>
+        /// <returns>The return value from the invoked method</returns>
+        public static PropertyInfo? GetIndexerByName(Type type, Type[] indexerTypes)
+        {
+            const string defaultFrameworkIndexerName = "Item";
+
+            var props = type.GetProperties(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance).ToArray();
+
+            var indexerInfo = props.FirstOrDefault(p => p.Name == defaultFrameworkIndexerName && ParametersEqual(p.GetIndexParameters(), indexerTypes));
+            return indexerInfo;
```

Comment:
Did you consider using using `type.GetProperty(indexerName, bindingFlags, binder: null, returnType: null, indexerTypes, modifiers: null)`?

Reply:
It would probably be best to use the same shadowing behavior that PropertyConstraint uses, eventually, but I don't think this is needed immediately.

Reply:
Yes, it was for multiple indexers that have similar parameters. It was returning the incorrect indexer (or throwing) as far as I can recall at this time. I'll have to have another look into this and confirm.

Reply:
So I've gone back to the tests. The `type.GetProperty` throws:
```
System.Reflection.AmbiguousMatchException
```
when used on a class that has something like:
```c#
public string this[int x] => // ...
public string this[string x] => // ...
public string this[int x, int y] => // ..
```

So in the event of multiple indexers in a class, the `GetProperties` needs to be iterated through.

Reply:
Ok, not sure what I did wrong, but this is working now..as you would expect.

Reply:
So this was interesting:
> It would probably be best to use the same shadowing behavior that PropertyConstraint uses

I had expected that it would also need this. However, I wrote some tests around shadow indexers and they behave a little differently. It looks to be pulling through the shadow indexers fine. 

If you could have a look at the tests in case I have missed something there.

Reply:
> Ok, not sure what I did wrong, but this is working now..as you would expect.

I think what is different is that this time you are using an overload of GetProperty that specifies the argument types.

Reply:
> If you could have a look at the tests in case I have missed something there.

This isn't quite what I was talking about. I was talking about mirroring what PropertyConstraint does, doing something analogous at each step where applicable. PropertyConstraint uses Reflect.GetUltimateShadowingProperty. Here are the tests that show why a method like that was needed: https://github.com/nunit/nunit/pull/2423/files#diff-246abb5484e1277f022d8d83ba9a6439dcc5d992f5d473f84e43c0a1c503de62

I don't think this is something we need to deal with in this PR, but I would expect that we would have the same thing happen with indexers that is shown in those tests. The fact that `PropertyConstraint` calls `Reflect.GetUltimateShadowingProperty` is a strong hint that `IndexerConstraint` should be calling a method called `Reflect.GetUltimateShadowingDefaultIndexer`.

Again, let's worry about this in a separate PR because this PR is complex enough, and I think the chance of hitting the bug is so tiny.

Reply:
Absolutely - I was expecting the need for `GetUltimateShadowingDefaultIndexer`.

So the added the tests to do the shadowing (not sure the tests are even correct) should have failed - they are similar to the property shadow tests. But I suspect the error you highlighted with the default/non-default indexer may be playing a role here.

If this does cause a problem once the other issue is resolved, then I'll remove the tests - but, agreed, not going to push deeply into this.